crates/ruma-macros/src/events/enums.rs

```diff
@@ -0,0 +1,574 @@
+        let stable_type = self.stable_type().ok_or_else(|| {
+            syn::Error::new(
+                Span::call_site(),
+                format!(
+                    "A matrix event must declare a well-known type that starts with `m.` \
+                     either as the main type or as an alias, or must declare the ident that \
+                     should be used if it is only an unstable type, found main type `{}`",
+                    self.ev_type
+                ),
+            )
+        })?;
+
+        Ok(m_prefix_name_to_type_name(&stable_type.source)
+            .expect("we already checked that the event type is stable"))
+    }
+}
+
+/// Common fields in event types.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum EventField {
+    OriginServerTs,
+    RoomId,
+    EventId,
+    Sender,
+}
+
+impl EventField {
+    /// All the variants of this enum
+    pub const ALL: &[Self] = &[Self::OriginServerTs, Self::RoomId, Self::EventId, Self::Sender];
+
+    /// Get the string representation of this field.
+    pub fn as_str(self) -> &'static str {
+        match self {
+            Self::OriginServerTs => "origin_server_ts",
+            Self::RoomId => "room_id",
+            Self::EventId => "event_id",
+            Self::Sender => "sender",
+        }
+    }
+
+    /// This field as an [`Ident`].
+    pub fn ident(self) -> Ident {
+        format_ident!("{}", self.as_str())
+    }
+
+    /// Whether this field is present in the given kind and variation.
+    pub fn is_present(self, kind: EventKind, var: EventVariation) -> bool {
+        match self {
+            Self::OriginServerTs | Self::EventId => {
+                kind.is_timeline()
+                    && matches!(
+                        var,
+                        EventVariation::None
+                            | EventVariation::Sync
+                            | EventVariation::Original
+                            | EventVariation::OriginalSync
+                            | EventVariation::Redacted
+                            | EventVariation::RedactedSync
+                    )
+            }
+            Self::RoomId => {
+                matches!(
+                    kind,
+                    EventKind::MessageLike
+                        | EventKind::State
+                        | EventKind::RoomRedaction
+                        | EventKind::EphemeralRoom
+                ) && matches!(
+                    var,
+                    EventVariation::None | EventVariation::Original | EventVariation::Redacted
+                )
+            }
+            Self::Sender => {
+                matches!(
+                    kind,
+                    EventKind::MessageLike
+                        | EventKind::State
+                        | EventKind::RoomRedaction
+                        | EventKind::ToDevice
+                ) && var != EventVariation::Initial
+            }
+        }
+    }
+
+    /// Get the return type of this field.
```

Comment:
```suggestion
    /// Get the type of this field.
```

I know this is a pre-existing mistake but this PR seems like a good place to fix it. Probably name the method `ty` instead of `return_type`?

Reply:
Changed.